BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m

```diff
@@ -167,50 +251,144 @@ - (void)deassociateSessionId:(NSString *)sessionId sessionTaskId:(NSUInteger)ses
+
+    if (backgroundSessionIdAndSessionTaskId != nil) {
+        NSString *backgroundSessionId = backgroundSessionIdAndSessionTaskId.backgroundSessionId;
+        NSUInteger sessionTaskId = backgroundSessionIdAndSessionTaskId.sessionTaskId;
+        NSURLSessionTask *task = [self existingBackgroundSessionTaskForSessionId:backgroundSessionId sessionTaskId:sessionTaskId];
+        returned = [[BOXBackgroundSessionIdAndTask alloc] initWithBackgroundSessionId:backgroundSessionId sessionTask:task];
+    }
+    return returned;
+}
+
+- (NSURLSessionTask *)backgroundTaskWithRequest:(NSURLRequest *)request orResumeDownloadData:(NSData *)resumeDownloadData orUploadFromFile:(NSURL *)uploadFromFileURL taskDelegate:(id <BOXURLSessionDownloadTaskDelegate>)taskDelegate userId:(NSString *)userId associateId:(NSString *)associateId
+{
+    NSError *error = nil;
+    NSURLSessionTask *sessionTask = nil;
+    BOXBackgroundSessionIdAndTask *task = [self existingBackgroundSessionTaskGivenUserId:userId associateId:associateId];
+
+    NSString *backgroundSessionId = nil;
+
+    if (task == nil) {
+        backgroundSessionId = backgroundSessionIdentifier;
+        sessionTask = uploadFromFileURL == nil ? (request != nil ? [self.backgroundSession downloadTaskWithRequest:request] : [self.backgroundSession downloadTaskWithResumeData:resumeDownloadData]) : [self.backgroundSession uploadTaskWithRequest:request fromFile:uploadFromFileURL];
+    } else {
+        backgroundSessionId = task.backgroundSessionId;
+        sessionTask = task.sessionTask;
+        if (backgroundSessionId != nil) {
+            NSURLSession *session = self.backgroundSessionIdToSession[backgroundSessionId];
+            if (resumeDownloadData != nil) {
+                sessionTask = [session downloadTaskWithResumeData:resumeDownloadData];
+            }
+        }
+    }
+
+    [self persistBackgroundSessionTaskWithSessionId:backgroundSessionId sessionTask:sessionTask taskDelegate:taskDelegate userId:userId associateId:associateId];
+    return sessionTask;
+}
+
+- (void)persistBackgroundSessionTaskWithSessionId:(NSString *)backgroundSessionId sessionTask:(NSURLSessionTask *)task taskDelegate:(id <BOXURLSessionDownloadTaskDelegate>)taskDelegate userId:(NSString *)userId associateId:(NSString *)associateId
+{
+    if (task == nil || taskDelegate == nil || userId == nil || associateId == nil) {
+        return;
+    }
+
+    [self associateSessionId:backgroundSessionId sessionTask:task withTaskDelegate:taskDelegate];
+
+    //cache background task if have not
+    NSError *error = nil;
+    BOOL success = [self.cacheClient cacheUserId:userId associateId:associateId backgroundSessionId:backgroundSessionId sessionTaskId:task.taskIdentifier error:&error];
+    BOXAssert(success, @"failed to cache background task %@", error);
```

Comment:
What do you think about bubbling up an error here instead of asserting? I wonder if it's still advantageous to give back the background task even if persisting it to disk somehow failed, but at least there's an error?